common/vm/vm.go

```diff
@@ -68,3 +88,35 @@ func createConnection(ip pulumi.StringInput, user string, env config.Environment
 
 	return connection.ToConnectionOutput(), nil
 }
+
+type VM interface {
```

Comment:
This interface sounds very limiting compared to what we already offer: `PackageManager`, `FileManager`, `DockerManager`. We're also probably very much interested in getting APIs like `PublicIP`, `PrivateIP` and maybe some others.
I am also not sure if the `VM` is a good conveyor for the `Environment` object.

I'm also not sure we should implement VM types here. The idea was more to have stuff like `aws.VM`, `azure.VM`, etc.. that implements that interface.

To maximize code re-utilisation, maybe we should use multiple interfaces, like:
```
type VM interface {
    OSVM,
    InfraVM,
}

type OSVM interface {
    PackageManager() command.PackageManager
    FileManager() ....
    DockerManager() ....
}

type InfraVM interface {
    PublicIP(),
    PrivateIP(),
}
```

Which would then allow to implement a `aws.VM` that takes an `OSVM` as parameter, and the provisioning API can easily return the `VM` type.

Reply:
> This interface sounds very limiting compared to what we already offer: PackageManager, FileManager, DockerManager

For `PackageManager`, I don't think it should be in a generic VM interface as packages installation is platform specific.  If you have a function that takes a `VM`, it would create the illusion that it would work for every VM.

`FileManager` could be provided with a VM, but `FileManager` is not cross platform. Also having a file manager that works on Docker and K8s would be useful. Tracked by.

`DockerManager` seems to be slower compared to ECS and also not working on Windows platform.

Reply:
> I am also not sure if the VM is a good conveyor for the Environment object.
Can you elaborate on this topic?

Reply:
> I'm also not sure we should implement VM types here. The idea was more to have stuff like aws.VM, azure.VM, etc.. that implements that interface.

I agree that having `aws.VM` and `azure.VM` makes sense, but the current code doesn't require it and introducing these concepts now or later cost the same. So let's introduce them when we have a need.

Reply:
As discussed in private message,  VM will contains `PackageManager`, `FileManager` and `DockerManager`